zipline/sources/data_frame_source.py

```diff
@@ -138,6 +145,12 @@ def raw_data_gen(self):
             df = self.data.major_xs(dt)
             for sid, series in df.iteritems():
                 if sid in self.sids:
+                    # Skip SIDs that can not be forward filled
```

Comment:
I don't think it's worth making this more complex. It'd require creation of a subclass, a call to an inherited method and an if-clause to remove 4 lines of simple code that are similar, but not identical.